### PR TITLE
Fast sync child trie support.

### PR DESCRIPTION
**Pull request by cheme**
This PR extend fast sync to support child trie.

When export reach a child root it includes it in the proof and switch to the child trie proof.
When child trie proof is added it resumes parent trie export.
Start key is now a list of nested trie storage key.

Currently only two level of trie are supported so it could also make sense to switch api from `start_key: Vec<Vec<u8>>` to `child_storage_key: Option<Vec<u8>>, start_key: Vec<u8>`.

https://github.com/paritytech/substrate/pull/9239/commits/0d08f53587a146c878644ba5aee4d4ab9dbb4d81 switches to compact proof usage (can be revert if needed).

### CODE REVIEW

File: primitives/state-machine/src/lib.rs
Comment (arkpar) [inline primitives/state-machine/src/lib.rs:747]: Would it make sense to identify subtree by storage root instead of a key paths?

File: client/network/src/protocol/sync/state.rs
Comment (arkpar) [inline client/network/src/protocol/sync/state.rs:153]: `response.entries` could be empty here?
Also, why `self.last_key.len() == 2`? Could use some explaining in the comment.

File: primitives/state-machine/src/lib.rs
Comment (arkpar) [inline primitives/state-machine/src/lib.rs:745]: Using name "State" here is a bit confusing. We already use the same term for the whole "State" as in the state of all storage for a block, including all subtrees. I'd suggest naming this something like "KeyValueStorageLevel"

File: primitives/state-machine/src/lib.rs
Comment (cheme) [reply primitives/state-machine/src/lib.rs:747]: Looks doable: instead of removing root from parent trie on response, we would register their storage key and build a payload that work for client.
It is a bit more work but I think it is a good idea.

File: client/network/src/protocol/sync/state.rs
Comment (cheme) [reply client/network/src/protocol/sync/state.rs:153]: It is empty when the whole change set is into a child trie: then we do not have the parent key in parent payload (already define in previous payload). I will add comment.

File: primitives/state-machine/src/lib.rs
Comment (cheme) [reply primitives/state-machine/src/lib.rs]: 4bd63f does it, it also avoid adding two child trie with same root in export.

**Review by arkpar — APPROVED** [review]

**Comment by arkpar** [comment]
@bkchr @andresilva Would be good to get reviews on this one in for warp sync as well, as it contains changes to the protocol.

**Comment by cheme** [comment]
Just need a second review. Note that it already contains the fix for https://github.com/paritytech/substrate/pull/9715, wish it was merge 2 month ago and did not have to investigate an refix it.

**Comment by cheme** [comment]
Yes looks needed. I was also thinking of moving the state sync message in their own protobuf file, but there is not much use in doing it except it separates things.

**Review by gavofyork — APPROVED** [review]

**Comment by gavofyork** [comment]
bot merge

**Pull request merged**
